### PR TITLE
Refactor ProviderBase, Test and Fetch classes to be generic

### PR DESCRIPTION
**Pull request by phamann**
### TL;DR
Refactors `ProviderBase`, `Test`, and `Fetch` classes to be generic and accept their configuration objects as type parameters. This allows provider authors to easily define and use their own custom types for each class without  sacrificing type safety. 

**🔎 Note for reviews:** as this has also regenerated documentation; it's easier to view the [commit](https://github.com/cbsinteractive/open-insights/pull/11/commits/e2a434f91e75197676aa6179e76bd6398a58d278) which has the code changes and not the full change set of files. 

### Why?
While implementing the `open-insights-provider-fastly` I found myself often battling with the global `TestConfiguration` and `SessionConfig` interfaces (which by-design were overly simple/generic to accomodate many use-cases) and regularly casting them to the Fastly providers internal types in order to access known properties on the objects to keep the compiler happy such as: `(testConfig as FastlyTaskData).customProp`etc.  This felt painful as a provider author, so I thought about how we could move the abstraction layer back to the core library - I _think_ TypeScript's [generics](https://www.typescriptlang.org/docs/handbook/generics.html) are the correct solution to this problem.
 
### How?
The `ProviderBase` class is now _generic_ and accepts two type properties, `SC` i.e. the old `SessionConfig` and `TC` i.e. `TestConfiguration`. So it now has the signature: 
```typescript
abstract class ProviderBase<SC, TC> implements Provider
``` 
This allows the consumer to supply the config types to use, while still allowing us to maintain generic business logic for all providers in the core library that operates on those types.

The `Provider` interface has been refactored to use the type-safe version of `any`, [`unknown`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type), wherever the SC or TC symbols are referenced. This allows you to have a list of different `Provider[]` interfaces and keep the compiler happy.

An example of how a provider author would use the class:
```typescript
import { ProviderBase } from "open-insights/src/lib/providerBase"

type CustomSessionConfig {
  sessionId: string,
  // other props ...
}

type CustomTestConfig {
  type: string,
  id: string
  url: string
  // other props ...
}

// Note how we now pass the config types to the ProviderBase signature.
class CustomProvider extends ProviderBase<CustomSessionConfig, CustomTestConfig> {
  // provider implementation ...
}
```

The same pattern can now be applied to the `Test` classes so they can have generic `TestConfiguration` supplied by the provider author:
```typescript
import Fetch from "open-insights/src/lib/fetch";

type CustomSessionConfig {
  sessionId: string,
  tasks: CustomTaskConfig[],
  // other props ...
}

type CustomTestConfig {
  type: string,
  id: string
  url: string
  // other props ...
}

class CustomProvider extends ProviderBase<CustomSessionConfig, CustomTestConfig> {
  //  ...

  function expandTasks(): Executable[] {
    return this.sessionConfig.tasks.map((t): Executeable => {
      // Note how we now pass the config type to the Fetch signature.
      return new Fetch<CustomTestConfig>(this, t)
    })
  }

  // ...
}
```

### Notes:
- To simplify the `init` logic I've also changed the signature of `SessionProcessFunc` to accept a list of executables `Executable[]` instead of the previous `SessionConfig[]` which no longer exists as a concrete type. However, I feel this is a cleaner implementation anyway. 
- I've had to refactor the tests to accommodate the new generics, so `UnitTestProvider` now has its own `UnitTestSessionConfig` and `UnitTestTestConfig` types.

### CODE REVIEW

**Comment by phamann** [comment]
Rebased with main to fix conflicts after #13 was merged.

**Comment by phamann** [comment]
Rebased with main to fix conflicts after #15 was merged.

**Review by jakewan — APPROVED** [review]
This is a great improvement. It helps me to begin thinking of code in hooks where we're having to cast to provider-refined types as a code smell, and hopefully, we can root these out in favor of generic implementations wherever we can, e.g. the return types of `getClientInfo` and `createFetchResult`.

The only odd thing I found was that when running jest tests in the CBSi provider, the Provider constructor somehow winds up invoking code in open-insights/src/init.ts, which I didn't expect. This isn't a blocker. It just suggests I'll have some further toubleshooting to do before we're fully back in sync with the open-insights main branch.

Really outstanding, @phamann. Thank you for this.
